insonmnia/worker/overseer.go

```diff
@@ -657,3 +659,88 @@ func (o *overseer) OnDealFinish(ctx context.Context, containerID string) error {
 func (o *overseer) Logs(ctx context.Context, id string, opts types.ContainerLogsOptions) (io.ReadCloser, error) {
 	return o.client.ContainerLogs(ctx, id, opts)
 }
+
+// prunedImage can stream pushed image with repository and tag data removed.
+type prunedImage struct {
+	image      *tar.Reader
+	writer     *tar.Writer
+	buf        *bytes.Buffer
+	currHeader *tar.Header
+}
+
+func newPrunedImage(image io.Reader) *prunedImage {
+	var buf = bytes.NewBuffer(nil)
+	return &prunedImage{
+		image:  tar.NewReader(image),
+		writer: tar.NewWriter(buf),
+		buf:    buf,
+	}
+}
+
+func (m *prunedImage) Read(p []byte) (n int, err error) {
+	m.buf.Reset()
+	if m.currHeader != nil {
+		// Read and send data for current file.
+		var in = make([]byte, len(p))
+		if n, err = m.image.Read(in); err == io.EOF {
+			// End of current file, switch to next header.
+			m.currHeader = nil
+		}
+		if _, err = m.writer.Write(in[:n]); err != nil {
+			return 0, err
+		}
+		return copy(p, m.buf.Bytes()), nil
+	} else {
+		// Try getting next header.
+		if m.currHeader, err = m.image.Next(); err == io.EOF {
+			// End of archive, write footer, exit with EOF.
+			if err = m.writer.Close(); err != nil {
+				return 0, fmt.Errorf("failed to write footer: %v", err)
+			}
+			return copy(p, m.buf.Bytes()), io.EOF
```

Comment:
looks useless, because buf was reset

Reply:
Close() flushes, writes padding and footer, buf is not empty after this call.